### ISSUE TITLE
mssql ConnectionError

### ISSUE THREAD

**Issue by sant123** [denoland/deno#20594]
Hi, I'm trying to run this code with mssql:

```ts
import sql from "npm:mssql@10.0.1";

const sqlConfig = {
  server: "localhost",
  user: "sa",
  password: "",
  database: "my_db",
  options: {
    trustServerCertificate: true,
  },
};

const pool = await sql.connect(sqlConfig);
const result = await pool.request().execute("spGetPurchases");

console.dir(result);

```

But I get `Uncaught (in promise) ConnectionError: Failed to connect to localhost:1433 in 15000ms`

![image](https://github.com/denoland/deno/assets/7959437/0b985b64-7653-4c06-b3c0-5d934d066aa5)

I decided to do the same with Node v18.16.0 and worked perfectly:

```js
const sql = require("mssql");

const sqlConfig = {
  server: "localhost",
  user: "sa",
  password: "",
  database: "my_db",
  options: {
    trustServerCertificate: true,
  },
};

(async () => {
  const pool = await sql.connect(sqlConfig);
  const result2 = await pool.request().execute("spGetPurchases");

  console.dir(result2);
})();

```
![image](https://github.com/denoland/deno/assets/7959437/bb59e95f-6844-46d1-b6e2-77fe60b68c27)

The SQL Server is running inside a Docker container:

![image](https://github.com/denoland/deno/assets/7959437/5e4d5034-b748-4bff-8a5d-5ea6c23c663a)

I'm using Deno latest version as of today v1.37.0

**Comment by sant123** [comment]
Alright I found some logs that may be helpful. I created a second project using Node so debugging is easier:

For Node v18.16.0, open node_modules/tedious/lib/connection.js#1040

In line 1041 add a console.log ->

![image](https://github.com/denoland/deno/assets/7959437/b19e33a3-c3e1-4333-84c2-0f74286a595b)

Back in Deno project, set `vendor: true` in deno.json and open node_modules/.deno/tedious@16.4.1/node_modules/tedious/lib/connection.js#1040

In line 1041 add a console.log ->

![image](https://github.com/denoland/deno/assets/7959437/4cf7c6a9-83ad-4e0f-aee5-43197cc6138b)

Node logs:

```
State change: Initialized -> Connecting
connected to localhost:1433
State change: Connecting -> SentPrelogin
State change: SentPrelogin -> SentTLSSSLNegotiation
TLS negotiated (ECDHE-RSA-AES128-GCM-SHA256, TLSv1.2)
State change: SentTLSSSLNegotiation -> SentLogin7WithStandardLogin
Packet size changed from 4096 to 4096
State change: SentLogin7WithStandardLogin -> LoggedInSendingInitialSql
State change: LoggedInSendingInitialSql -> LoggedIn
State change: LoggedIn -> Final
connection to localhost:1433 closed
State is already Final
State change: Initialized -> Connecting
connected to localhost:1433
State change: Connecting -> SentPrelogin
State change: SentPrelogin -> SentTLSSSLNegotiation
TLS negotiated (ECDHE-RSA-AES128-GCM-SHA256, TLSv1.2)
State change: SentTLSSSLNegotiation -> SentLogin7WithStandardLogin
Packet size changed from 4096 to 4096
State change: SentLogin7WithStandardLogin -> LoggedInSendingInitialSql
State change: LoggedInSendingInitialSql -> LoggedIn
State change: LoggedIn -> SentClientRequest
State change: SentClientRequest -> LoggedIn
State change: LoggedIn -> SentClientRequest
State change: SentClientRequest -> LoggedIn
```

Deno logs:

```
State change: Initialized -> Connecting
connected to localhost:1433
State change: Connecting -> SentPrelogin
State change: SentPrelogin -> SentTLSSSLNegotiation
Failed to connect to localhost:1433 in 15000ms
State change: SentTLSSSLNegotiation -> Final
connection to localhost:1433 closed
State is already Final
```

It seems there is a issue when attempting to negotiate the TLS certificate.

**Comment by massalinux** [comment]
hi @sant123 , i got the same issue, while is not a fix i managed to proceed with   ` encrypt: false`

**Comment by mmastrac** [comment]
Could you run:

`openssl s_client -connect localhost:1433`

... and paste the output here? It may be that the underlying cipher supported by the DB engine is not supported by our TLS code.

**Comment by massalinux** [comment]
`8085DC0201000000:error:8000003D:system library:BIO_connect:Connection refused:crypto/bio/bio_sock2.c:114:calling connect()
8085DC0201000000:error:10000067:BIO routines:BIO_connect:connect error:crypto/bio/bio_sock2.c:116:
connect:errno=61
`

**Comment by mmastrac** [comment]
It looks like the TLS cipher should be supported -- I will make sure that we correctly enable it for our tls library.

**Comment by mmastrac** [comment]
@massalinux Could you run this from Deno 1.38.2, but using the address of your database?

```
let hostname = 'HOSTNAME';
let port = PORT_NUMBER;
let conn = await Deno.connectTls({hostname, port});
console.log(await conn.handshake();)
```

**Comment by massalinux** [comment]
@mmastrac here it is

`error: Uncaught (in promise) ConnectionRefused: Connection refused (os error 61)
let conn = await Deno.connectTls({hostname, port});
           ^
    at async Object.connectTls (ext:deno_net/02_tls.js:36:51)
    at async file:///Users/martino/d/varie/variedeno/main.ts:3:12`

**Comment by mmastrac** [comment]
Hmm. It looks like the server is not running or is not accessible. Is the docker container running? Can you access it any other way?

You maybe need to use 0.0.0.0 as a hostname

**Comment by massalinux** [comment]
Hi Matt, actually i can connect to that address at that port to SQLServer via mssql (with the encrypt: false option),  do you want me to run additional tests?

**Comment by sant123** [comment]
> `openssl s_client -connect localhost:1433`

```text
CONNECTED(00000003)
009E1113B37F0000:error:0A000126:SSL routines:ssl3_read_n:unexpected eof while reading:ssl/record/rec_layer_s3.c:303:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 302 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
This TLS version forbids renegotiation.
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

Running this code with Deno v1.38.3

```js
const hostname = "localhost";
const port = 1433;
const conn = await Deno.connectTls({ hostname, port });
console.log(await conn.handshake());
```

![image](https://github.com/denoland/deno/assets/7959437/374790b8-9c59-460a-addf-fd0afffa8bbb)

My Docker instance:

![image](https://github.com/denoland/deno/assets/7959437/85904eb7-af44-4b68-91ef-8ea44ce3a5e9)

**Comment by sant123** [comment]
These are the logs from the container:

```text
2023-11-27 17:08:13.40 Logon       A valid TLS certificate is not configured to accept strict (TDS 8.0 and above) connections. The connection has been closed.
2023-11-27 17:08:28.72 Logon       Error: 17821, Severity: 20, State: 1.
2023-11-27 17:08:28.72 Logon       A valid TLS certificate is not configured to accept strict (TDS 8.0 and above) connections. The connection has been closed.
```

**Comment by jtomaszewski** [comment]
I have the same troubles as in https://github.com/denoland/deno/issues/20594#issuecomment-1730047671 while connecting to an Azure MS SQL server using mssql library .

In node.js, I can connect easily. In deno, when running 

```ts
import mssql from "npm:mssql";
await mssql.connect("Server=abcdef-dev-sql-db.123456.database.windows.net,1433;Database=somedb;User Id=someuser;Password=somepassword"); # I've replaced the actual credentials here with fake ones ofc
```

I get this output:

```
State change: Initialized -> Connecting
connected to abcdef-dev-sql-db.123456.database.windows.net:1433
State change: Connecting -> SentPrelogin
State change: SentPrelogin -> SentTLSSSLNegotiation
connection to abcdef-dev-sql-db.123456.database.windows.net:1433 closed
State change: SentTLSSSLNegotiation -> Final
Connection lost - read ECONNRESET
Uncaught ConnectionError: Connection lost - read ECONNRESET
    at file:///Users/jacek/src/myevaluations/windmill/node_modules/.deno/mssql@10.0.1/node_modules/mssql/lib/tedious/connection-pool.js:85:17
    at Connection.onError (file:///Users/jacek/src/myevaluations/windmill/node_modules/.deno/tedious@16.6.1/node_modules/tedious/lib/connection.js:842:9)
    at Object.onceWrapper (ext:deno_node/_events.mjs:508:26)
    at Connection.emit (ext:deno_node/_events.mjs:395:35)
    at Connection.emit (file:///Users/jacek/src/myevaluations/windmill/node_modules/.deno/tedious@16.6.1/node_modules/tedious/lib/connection.js:959:18)
    at Connection.socketError (file:///Users/jacek/src/myevaluations/windmill/node_modules/.deno/tedious@16.6.1/node_modules/tedious/lib/connection.js:1345:12)
    at Socket.<anonymous> (file:///Users/jacek/src/myevaluations/windmill/node_modules/.deno/tedious@16.6.1/node_modules/tedious/lib/connection.js:1048:12)
    at Socket.emit (ext:deno_node/_stream.mjs:1854:9)
    at emitErrorNT (ext:deno_node/_stream.mjs:1572:13)
    at emitErrorCloseNT (ext:deno_node/_stream.mjs:1544:7)
```

`deno --version` output:

```
deno 1.38.5 (release, aarch64-apple-darwin)
v8 12.0.267.1
typescript 5.2.2
```

Testing `Deno.connectTls`:

input:
```
let hostname = 'abcdef-dev-sql-db.123456.database.windows.net';
let port = 1433;
let conn = await Deno.connectTls({hostname, port});
console.log(await conn.handshake())
```

output:
```
[Object: null prototype] { alpnProtocol: null }
```

Output for `openssl s_client -connect abcdef-dev-sql-db.123456.database.windows.net:1433`:

```
depth=2 C=US, O=DigiCert Inc, OU=www.digicert.com, CN=DigiCert Global Root G2
verify return:1
depth=1 C=US, O=Microsoft Corporation, CN=Microsoft Azure TLS Issuing CA 05
verify return:1
depth=0 C=US, ST=WA, L=Redmond, O=Microsoft Corporation, CN=502330f57f6f.database.windows.net
verify return:1
Connecting to 10.10.1.254
CONNECTED(00000005)
---
Certificate chain
 0 s:C=US, ST=WA, L=Redmond, O=Microsoft Corporation, CN=502330f57f6f.database.windows.net
   i:C=US, O=Microsoft Corporation, CN=Microsoft Azure TLS Issuing CA 05
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA384
   v:NotBefore: Nov 16 19:33:25 2023 GMT; NotAfter: Jun 27 23:59:59 2024 GMT
 1 s:C=US, O=Microsoft Corporation, CN=Microsoft Azure TLS Issuing CA 05
   i:C=US, O=DigiCert Inc, OU=www.digicert.com, CN=DigiCert Global Root G2
   a:PKEY: rsaEncryption, 4096 (bit); sigalg: RSA-SHA384
   v:NotBefore: Jul 29 12:30:00 2020 GMT; NotAfter: Jun 27 23:59:59 2024 GMT
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIJLjCCBxa....
-----END CERTIFICATE-----
subject=C=US, ST=WA, L=Redmond, O=Microsoft Corporation, CN=502330f57f6f.database.windows.net
issuer=C=US, O=Microsoft Corporation, CN=Microsoft Azure TLS Issuing CA 05
---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: ECDH, secp384r1, 384 bits
---
SSL handshake has read 4534 bytes and written 865 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 2048 bit
This TLS version forbids renegotiation.
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
read:errno=54
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: 3DC1...
    Session-ID-ctx: 
    Resumption PSK: C2B4...
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 36000 (seconds)
    TLS session ticket:
    0000 - d1 ...
    0010 - 2c ...

    Start Time: 1702287665
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK
```

@mmastrac let me know if I can help anyhow differently in this.

Unfortunately this currently blocks me in using mssql & deno for my project, as I need it to connect to Azure MS SQL :(

**Comment by sant123** [comment]
Using Deno v1.40.1 now logs the following warnings:

```
warning: Use of deprecated "Deno.TcpConn.rid" API. This API will be removed in Deno 2.

Stack trace:
  at ext:deno_node/internal_binding/tcp_wrap.ts:295:14

hint: Use `Deno.TcpConn` instance methods instead.
hint: It appears this API is used by a remote dependency. Try upgrading to the latest version of that dependency.

warning: Use of deprecated "Deno.TcpConn.rid" API. This API will be removed in Deno 2.

hint: Use `Deno.TcpConn` instance methods instead.

warning: Use of deprecated "Deno.TcpConn.rid" API. This API will be removed in Deno 2.

hint: Use `Deno.TcpConn` instance methods instead.

warning: Use of deprecated "Deno.TcpConn.rid" API. This API will be removed in Deno 2.

hint: Use `Deno.TcpConn` instance methods instead.

error: Uncaught (in promise) ConnectionError: Failed to connect to localhost:1433 in 15000ms
    at file:///home/sant821/.cache/deno/npm/registry.npmjs.org/mssql/10.0.2/lib/tedious/connection-pool.js:85:17
    at Connection.onConnect (file:///home/sant821/.cache/deno/npm/registry.npmjs.org/tedious/16.6.1/lib/connection.js:838:9)
    at Object.onceWrapper (ext:deno_node/_events.mjs:508:26)
    at Connection.emit (ext:deno_node/_events.mjs:383:28)
    at Connection.emit (file:///home/sant821/.cache/deno/npm/registry.npmjs.org/tedious/16.6.1/lib/connection.js:959:18)
    at Connection.connectTimeout (file:///home/sant821/.cache/deno/npm/registry.npmjs.org/tedious/16.6.1/lib/connection.js:1209:10)
    at Timeout._onTimeout (file:///home/sant821/.cache/deno/npm/registry.npmjs.org/tedious/16.6.1/lib/connection.js:1154:12)
    at cb (ext:deno_node/internal/timers.mjs:63:31)
    at Object.action (ext:deno_web/02_timers.js:154:11)
    at handleTimerMacrotask (ext:deno_web/02_timers.js:68:10)
```

**Comment by pcc-huangm** [comment]
> I have the same troubles as in [#20594 (comment)](https://github.com/denoland/deno/issues/20594#issuecomment-1730047671) while connecting to an Azure MS SQL server using mssql library .
> 
> In node.js, I can connect easily. In deno, when running
> 
> ```ts
> import mssql from "npm:mssql";
> await mssql.connect("Server=abcdef-dev-sql-db.123456.database.windows.net,1433;Database=somedb;User Id=someuser;Password=somepassword"); # I've replaced the actual credentials here with fake ones ofc
> ```
> 
> I get this output:
> 
> ```
> State change: Initialized -> Connecting
> connected to abcdef-dev-sql-db.123456.database.windows.net:1433
> State change: Connecting -> SentPrelogin
> State change: SentPrelogin -> SentTLSSSLNegotiation
> connection to abcdef-dev-sql-db.123456.database.windows.net:1433 closed
> State change: SentTLSSSLNegotiation -> Final
> Connection lost - read ECONNRESET
> Uncaught ConnectionError: Connection lost - read ECONNRESET
>     at file:///Users/jacek/src/myevaluations/windmill/node_modules/.deno/mssql@10.0.1/node_modules/mssql/lib/tedious/connection-pool.js:85:17
>     at Connection.onError (file:///Users/jacek/src/myevaluations/windmill/node_modules/.deno/tedious@16.6.1/node_modules/tedious/lib/connection.js:842:9)
>     at Object.onceWrapper (ext:deno_node/_events.mjs:508:26)
>     at Connection.emit (ext:deno_node/_events.mjs:395:35)
>     at Connection.emit (file:///Users/jacek/src/myevaluations/windmill/node_modules/.deno/tedious@16.6.1/node_modules/tedious/lib/connection.js:959:18)
>     at Connection.socketError (file:///Users/jacek/src/myevaluations/windmill/node_modules/.deno/tedious@16.6.1/node_modules/tedious/lib/connection.js:1345:12)
>     at Socket.<anonymous> (file:///Users/jacek/src/myevaluations/windmill/node_modules/.deno/tedious@16.6.1/node_modules/tedious/lib/connection.js:1048:12)
>     at Socket.emit (ext:deno_node/_stream.mjs:1854:9)
>     at emitErrorNT (ext:deno_node/_stream.mjs:1572:13)
>     at emitErrorCloseNT (ext:deno_node/_stream.mjs:1544:7)
> ```
> 
> `deno --version` output:
> 
> ```
> deno 1.38.5 (release, aarch64-apple-darwin)
> v8 12.0.267.1
> typescript 5.2.2
> ```
> 
> Testing `Deno.connectTls`:
> 
> input:
> 
> ```
> let hostname = 'abcdef-dev-sql-db.123456.database.windows.net';
> let port = 1433;
> let conn = await Deno.connectTls({hostname, port});
> console.log(await conn.handshake())
> ```
> 
> output:
> 
> ```
> [Object: null prototype] { alpnProtocol: null }
> ```
> 
> Output for `openssl s_client -connect abcdef-dev-sql-db.123456.database.windows.net:1433`:
> 
> ```
> depth=2 C=US, O=DigiCert Inc, OU=www.digicert.com, CN=DigiCert Global Root G2
> verify return:1
> depth=1 C=US, O=Microsoft Corporation, CN=Microsoft Azure TLS Issuing CA 05
> verify return:1
> depth=0 C=US, ST=WA, L=Redmond, O=Microsoft Corporation, CN=502330f57f6f.database.windows.net
> verify return:1
> Connecting to 10.10.1.254
> CONNECTED(00000005)
> ---
> Certificate chain
>  0 s:C=US, ST=WA, L=Redmond, O=Microsoft Corporation, CN=502330f57f6f.database.windows.net
>    i:C=US, O=Microsoft Corporation, CN=Microsoft Azure TLS Issuing CA 05
>    a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA384
>    v:NotBefore: Nov 16 19:33:25 2023 GMT; NotAfter: Jun 27 23:59:59 2024 GMT
>  1 s:C=US, O=Microsoft Corporation, CN=Microsoft Azure TLS Issuing CA 05
>    i:C=US, O=DigiCert Inc, OU=www.digicert.com, CN=DigiCert Global Root G2
>    a:PKEY: rsaEncryption, 4096 (bit); sigalg: RSA-SHA384
>    v:NotBefore: Jul 29 12:30:00 2020 GMT; NotAfter: Jun 27 23:59:59 2024 GMT
> ---
> Server certificate
> -----BEGIN CERTIFICATE-----
> MIIJLjCCBxa....
> -----END CERTIFICATE-----
> subject=C=US, ST=WA, L=Redmond, O=Microsoft Corporation, CN=502330f57f6f.database.windows.net
> issuer=C=US, O=Microsoft Corporation, CN=Microsoft Azure TLS Issuing CA 05
> ---
> No client certificate CA names sent
> Peer signing digest: SHA256
> Peer signature type: RSA-PSS
> Server Temp Key: ECDH, secp384r1, 384 bits
> ---
> SSL handshake has read 4534 bytes and written 865 bytes
> Verification: OK
> ---
> New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
> Server public key is 2048 bit
> This TLS version forbids renegotiation.
> Compression: NONE
> Expansion: NONE
> No ALPN negotiated
> Early data was not sent
> Verify return code: 0 (ok)
> ---
> read:errno=54
> ---
> Post-Handshake New Session Ticket arrived:
> SSL-Session:
>     Protocol  : TLSv1.3
>     Cipher    : TLS_AES_256_GCM_SHA384
>     Session-ID: 3DC1...
>     Session-ID-ctx: 
>     Resumption PSK: C2B4...
>     PSK identity: None
>     PSK identity hint: None
>     SRP username: None
>     TLS session ticket lifetime hint: 36000 (seconds)
>     TLS session ticket:
>     0000 - d1 ...
>     0010 - 2c ...
> 
>     Start Time: 1702287665
>     Timeout   : 7200 (sec)
>     Verify return code: 0 (ok)
>     Extended master secret: no
>     Max Early Data: 0
> ---
> read R BLOCK
> ```
> 
> @mmastrac let me know if I can help anyhow differently in this.
> 
> Unfortunately this currently blocks me in using mssql & deno for my project, as I need it to connect to Azure MS SQL :(
> 
> P.S. My mssql server requires VPN connection. I've got enabled it globally on my OS X. I can connect to mssql server using node.js and DataGrip easily (without the need of preconfiguring any certificates in the connection settings). I tried adding `DENO_TLS_CA_STORE=system` to the `deno` but it doesn't help either.

Having the exact issue, unable to connect to Azure MSSQL dbs.

**Comment by KnorpelSenf** [comment]
@nathanwhit are you planning on working on this in the near future?

**Comment by egous-bunnings** [comment]
Just like many others I too am blocked.
The workaround of `encrypt: false` is not possible within the Azure SQL context [as they enforce SSL
](https://learn.microsoft.com/en-us/azure/azure-sql/database/security-overview?view=azuresql#transport-layer-security-encryption-in-transit)

Unfortunately this is the bug that will force me to ditch Deno as not other possible workaround exists.

**Comment by kbrown-remed-com** [comment]
Like many others, I have been waiting for a fix for this.  I was hoping that this would work in 2.0, but I'm still getting the exact same errors as I did in 1.4x.  I would love to begin to use Deno, but this bug is preventing me from switching.

**Comment by mderazon** [comment]
Trying to convert a Node project to Deno and hitting the same issue. As a last measure, has anyone found a different module that doesn't have this problem ? Or is it something fundumental with the way Deno connects to MSSQL server ?

**Comment by egous-bunnings** [comment]
@mderazon I couldn't find a workable solution/workaround to this issue and was forced to port out.

**Comment by v1d3rm3** [comment]
I believe that adoption of Deno depends of creators officially support ways to connect to database in platform.

**Comment by QimatLuo** [comment]
> Just like many others I too am blocked. The workaround of `encrypt: false` is not possible within the Azure SQL context [as they enforce SSL ](https://learn.microsoft.com/en-us/azure/azure-sql/database/security-overview?view=azuresql#transport-layer-security-encryption-in-transit)
> 
> Unfortunately this is the bug that will force me to ditch Deno as no other possible workaround exists.

Works for me for development purpose, hope it can fix before our project needs to go to production.

**Comment by tjcrowder** [comment]
@littledivy - Any chance of an update on this with an ETA? I'm surprised this has sat so long. Deno generally seems to move faster than this... TIA!

**Comment by kbrown-remed-com** [comment]
@bartlomieju Can you please provide any kind of update on where this issue sits?  It's been added and removed from roadmaps and milestones for a year now.  For myself and many others, this is the single barrier preventing us from using Deno over Node.

**Comment by arthur-ver** [comment]
I’m also waiting on a fix for this issue for months now, since I can’t use `encrypt: false` in production.

**Comment by sant123** [comment]
> The issue you referenced, [denoland/deno#20594](https://github.com/denoland/deno/issues/20594), points to a problem where Deno fails to complete the TLS handshake when trying to connect to SQL Server with SSL/TLS enabled. This is indeed a strange behavior, especially since Deno is generally built with a focus on security and includes built-in support for handling TLS/SSL in other contexts (e.g., HTTP/HTTPS connections).

> Here are some possible underlying causes for this issue:
> 1. Incomplete or Incorrect TLS Implementation in the Runtime:

> The Deno runtime itself handles a variety of low-level operations, including TLS handshake processes. If the TLS implementation within Deno is incomplete or has issues specifically related to database protocols (e.g., TDS for SQL Server), this could explain why the handshake is not happening. Unlike a regular HTTPS request, database communication protocols like PostgreSQL and SQL Server involve more specialized TLS handling, which might expose gaps or bugs in Deno's networking stack.
> 2. Database-Specific SSL Configuration:

> SQL Server and PostgreSQL might have unique requirements when negotiating TLS connections, such as custom certificate handling or specific protocol versions that Deno might not fully support. There could be incompatibilities between the way Deno establishes SSL connections and how the database expects them to be initiated.

> For example, SQL Server requires certain extensions during the handshake that might not be correctly managed by Deno. If Deno's native support for SSL/TLS is not aligned with the database’s expectations (for example, related to certificate exchange, ciphers, or protocol version), the handshake would fail before it starts.
> 3. Certificate Validation Issues:

> TLS handshakes require proper validation of certificates from both sides (client and server). If Deno is not properly handling certificate chains, missing CA certificates, or mishandling self-signed certificates, it could prevent the handshake from happening.

> In this case, if Deno isn't sending the correct client certificate or failing to validate the server's certificate properly, the handshake could fail immediately. This might result in the behavior seen in the issue, where no handshake is made.
> 4. Library/Dependency Incompatibility:

> Deno uses Rust-based libraries under the hood for many of its core functionalities, including its networking stack. If there's an issue or bug in one of these underlying libraries, such as rustls (used by Deno for handling TLS), that could be affecting the ability to handle secure database connections.

> A mismatch between Deno's networking stack and the specific requirements for initiating SQL Server TLS connections could be at play. It’s possible that the version of rustls or another related library might have known issues with database TLS connections.
> 5. Platform-Specific Behavior:

> The issue could also be specific to certain platforms (e.g., Linux vs. Windows) or network configurations. Some users report that disabling SSL on the database allows the connection to work, which suggests that this is purely a TLS issue. Deno may behave differently in different environments, especially when interacting with certain database drivers, SSL configurations, or network setups.
> Conclusion:

> This appears to be an edge case or a limitation in Deno’s current TLS implementation, especially concerning database drivers like those for SQL Server and PostgreSQL. Given that Deno is relatively new compared to other runtimes, the networking and database driver compatibility might still be maturing. The failure in the TLS handshake indicates that the runtime might not be handling something critical in the SSL negotiation process (such as certificate validation, protocol version mismatch, or cipher suite selection).

> The issue would likely need further investigation by the Deno core team to pinpoint whether it's related to rustls or another lower-level implementation detail. For now, users facing this problem are left with workarounds like disabling SSL, which isn’t ideal in production environments.

**Comment by oscar6echo** [comment]
I have a similar issue with a postgresql db.  
Impossible to connect while it works fine in node and python.  
I get 

Cf. #28551 

I find it suprising that such a foundational feature seems buggy in Deno, whose team is often super reactive.  
It makes is impossible to consider Deno in a microservice context - at least those using a db.  
Unless this bug is very difficult to solve, this blind spot really is detrimental to Deno adoption in a corp prod env.

**Comment by garytube** [comment]
Same here, I would love to use Deno in enterprise land where mssql ist common

**Comment by sant123** [comment]
Gentle bump on this please 🙏🏼 @ry @bartlomieju @satyarohith

**Comment by oscar6echo** [comment]
SSL errors may have several causes - I suspect.  
But in my case - failure to connect to an Azure PostgreSQL - I fixed it by investigating the certificate on the DB side.  
Cf. https://github.com/denoland/deno/issues/28903#issuecomment-2809728945

In the case of mssql probably the only change is in the openssl command - to be checked.  

```sh
echo | openssl s_client -connect your_mssql_server:1433
```

**Comment by satyarohith** [comment]
The main problem blocking this issue is mssql terminates TLS on a duplex stream. But Deno internals only support terminating TLS on a TCP stream. We need infra to terminate TLS on any AsyncWrite+AsyncRead . To achieve that we tried to implement a JsStream struct that provides interoperability between rust and JS but it has been unsuccessful due to node internals relying heavily on the duplex being a fake net.Socket. So in Deno we need to essentially convert a duplex to a fake net.Socket that can be read/writen from rust for TLS termination.

Work by Divy/Satya

**Comment by dannybunschoten** [comment]
I can confirm that using `encrypt: false` also worked for me. The issue indeed seems related to authentication. The same connection config did work on nodejs for me.

**Comment by arthur-ver** [comment]
The issue still persists for me on the latest Deno version `2.4.5`. Still only works with `encrypt: false`.

**Comment by bartlomieju** [comment]
@arthur-ver could you please open a new issue with your reproduction code? Thanks

**Comment by tjcrowder** [comment]
@bartlomieju - Would the fix even be in v2.4.5? The PR was just on Aug 20th and approved on Aug 28th. v2.4.5 was released on Aug 21st.

@arthur-ver Can you try Deno Canary? `deno upgrade canary` will upgrade Deno to the canary build (basically nightly) which should have the fix, then just `deno upgrade` will take you back to 2.4.5 when you've finished your test.

**Comment by bartlomieju** [comment]
Oh, you're right 🤦 the fix is only available on `canary`.

**Comment by arthur-ver** [comment]
@tjcrowder `canary` works!

@bartlomieju my bad, I thought the fix was released with `2.4.5`

**Comment by tjcrowder** [comment]
@littledivy - **Thank you** for your work on this! This was a roadblock for me on a project a while back. Really glad it won't be next time, thanks to you!

**Comment by jhallam3** [comment]
Hi I'm trying to use this in supabase edge functions is this possible?
